reactos/drivers/usb/usbhub_new/usbhub.c

```diff
@@ -0,0 +1,5143 @@
+    //       Irp);
+
+    IoSkipCurrentIrpStackLocation(Irp);
+    return IoCallDriver(DeviceObject, Irp);
+}
+
+NTSTATUS
+NTAPI
+USBH_SyncIrpComplete(IN PDEVICE_OBJECT DeviceObject,
+                     IN PIRP Irp,
+                     IN PVOID Context)
+{
+    PUSBHUB_URB_TIMEOUT_CONTEXT HubTimeoutContext;
+    KIRQL OldIrql;
+    BOOLEAN TimerCancelled;
+
+    DPRINT("USBH_SyncIrpComplete: ... \n");
+
+    HubTimeoutContext = Context;
+
+    KeAcquireSpinLock(&HubTimeoutContext->UrbTimeoutSpinLock, &OldIrql);
+    HubTimeoutContext->IsNormalCompleted = TRUE;
+    TimerCancelled = KeCancelTimer(&HubTimeoutContext->UrbTimeoutTimer);
+    KeReleaseSpinLock(&HubTimeoutContext->UrbTimeoutSpinLock, OldIrql);
+
+    if (TimerCancelled)
+    {
+        KeSetEvent(&HubTimeoutContext->UrbTimeoutEvent,
+                   EVENT_INCREMENT,
+                   FALSE);
+    }
+
+    return STATUS_SUCCESS;
+}
+
+BOOLEAN
+NTAPI
+IsBitSet(IN ULONG_PTR BitMapAddress,
+         IN ULONG Bit)
+{
+    BOOLEAN IsSet;
+
+    IsSet = (*(PUCHAR)(BitMapAddress + (Bit >> 3)) & (1 << (Bit & 7))) != 0;
+    DPRINT("IsBitSet: Bit - %p, IsSet - %x\n", Bit, IsSet);
+    return IsSet;
+}
+
+PUSBHUB_PORT_PDO_EXTENSION
+NTAPI
+PdoExt(IN PDEVICE_OBJECT DeviceObject)
+{
+    PVOID PdoExtension;
+
+    DPRINT("PdoExt: DeviceObject - %p\n", DeviceObject);
+
+    if (DeviceObject)
+    {
+        PdoExtension = DeviceObject->DeviceExtension;
+    }
+    else
+    {
+        PdoExtension = (PVOID)-1;
+    }
+
+    return (PUSBHUB_PORT_PDO_EXTENSION)PdoExtension;
+}
+
+NTSTATUS
+NTAPI
+USBH_WriteFailReasonID(IN PDEVICE_OBJECT DeviceObject,
+                       IN ULONG Data)
+{
+    NTSTATUS Status;
+    WCHAR SourceString[64];
+    HANDLE KeyHandle;
+    UNICODE_STRING DestinationString;
+
+    DPRINT("USBH_WriteFailReason: ID - %x\n", Data);
+
+    swprintf(SourceString, L"FailReasonID");
+
+    Status = IoOpenDeviceRegistryKey(DeviceObject,
+                                     PLUGPLAY_REGKEY_DEVICE,
+                                     STANDARD_RIGHTS_ALL,
+                                     &KeyHandle);
+
+    if (NT_SUCCESS(Status))
+    {
+        RtlInitUnicodeString(&DestinationString, SourceString);
+
+        return ZwSetValueKey(KeyHandle,
+                             &DestinationString,
+                             0,
+                             REG_DWORD,
+                             &Data,
+                             sizeof(Data));
+
+        ZwClose(KeyHandle);
+    }
+
+    return Status;
+}
+
+VOID
+NTAPI
+USBH_UrbTimeoutDPC(IN PKDPC Dpc,
+                   IN PVOID DeferredContext,
+                   IN PVOID SystemArgument1,
+                   IN PVOID SystemArgument2)
+{
+    PUSBHUB_URB_TIMEOUT_CONTEXT HubTimeoutContext;
+    KIRQL OldIrql;
+    BOOL IsCompleted;
+
+    DPRINT("USBH_TimeoutDPC ... \n");
+
+    HubTimeoutContext = DeferredContext;
+
+    KeAcquireSpinLock(&HubTimeoutContext->UrbTimeoutSpinLock, &OldIrql);
```

Comment:
usbhub.c:186: Concurrent access shouldn't be a problem. Although I suppose it could be on some exotic architecture.
Anyway, it seems okay to leave as-is.